### PR TITLE
[Feature][Bugfix] Support draft model tp any of speculative decode

### PR DESCRIPTION
**Pull request by stormchasingg**
## Purpose
We met a problem that adding `'draft_tensor_parallel_size': 1` or `'draft_tensor_parallel_size': 2` in server script, the draft model tensor parallel size is still same as the main model.
```
Qwen3ForCausalLM(
  (model): Qwen3Model(
    (embed_tokens): VocabParallelEmbedding(num_embeddings=18992, embedding_dim=5120, org_vocab_size=151936, num_embeddings_padded=151936, tp_size=8)
    (layers): ModuleList(
      (0-63): 64 x Qwen3DecoderLayer(
        (self_attn): Qwen3Attention(
          (qkv_proj): QKVParallelLinear(in_features=5120, output_features=1280, bias=False, tp_size=8, gather_output=False)
          (o_proj): RowParallelLinear(in_features=1024, output_features=5120, bias=False, tp_size=8, reduce_results=True)
          (rotary_emb): YaRNScalingRotaryEmbedding(head_size=128, rotary_dim=128, max_position_embeddings=40960, base=1000000, is_neox_style=True)
          (attn): Attention(head_size=128, num_heads=8, num_kv_heads=1, scale=0.08838834764831845, backend=FlashAttentionImpl)
          (q_norm): RMSNorm(hidden_size=128, eps=1e-06)
          (k_norm): RMSNorm(hidden_size=128, eps=1e-06)
        )
        (mlp): Qwen2MLP(
          (gate_up_proj): MergedColumnParallelLinear(in_features=5120, output_features=6400, bias=False, tp_size=8, gather_output=False)
          (down_proj): RowParallelLinear(in_features=3200, output_features=5120, bias=False, tp_size=8, reduce_results=True)
          (act_fn): SiluAndMul()
        )
        (input_layernorm): RMSNorm(hidden_size=5120, eps=1e-06)
        (post_attention_layernorm): RMSNorm(hidden_size=5120, eps=1e-06)
      )
    )
    (norm): RMSNorm(hidden_size=5120, eps=1e-06)
  )
  (lm_head): ParallelLMHead(num_embeddings=18992, embedding_dim=5120, org_vocab_size=151936, num_embeddings_padded=151936, tp_size=8)
  (logits_processor): LogitsProcessor(vocab_size=151936, org_vocab_size=151936, scale=1.0, logits_as_input=False)
)

EagleQwen3ForCausalLM(
  (model): Qwen3Model(
    (embed_tokens): VocabParallelEmbedding(num_embeddings=18992, embedding_dim=5120, org_vocab_size=151936, num_embeddings_padded=151936, tp_size=8)
    (layers): ModuleList(
      (0): Qwen3EalgeDecoderLayer(
        (self_attn): Qwen3Attention(
          (qkv_proj): QKVParallelLinear(in_features=5120, output_features=1280, bias=False, tp_size=8, gather_output=False)
          (o_proj): RowParallelLinear(in_features=1024, output_features=5120, bias=False, tp_size=8, reduce_results=True)
          (rotary_emb): YaRNScalingRotaryEmbedding(head_size=128, rotary_dim=128, max_position_embeddings=40960, base=1000000, is_neox_style=True)
          (attn): Attention(head_size=128, num_heads=8, num_kv_heads=1, scale=0.08838834764831845, backend=FlashAttentionImpl)
          (q_norm): RMSNorm(hidden_size=128, eps=1e-06)
          (k_norm): RMSNorm(hidden_size=128, eps=1e-06)
        )
        (mlp): Qwen3MLP(
          (gate_up_proj): MergedColumnParallelLinear(in_features=5120, output_features=6400, bias=False, tp_size=8, gather_output=False)
          (down_proj): RowParallelLinear(in_features=3200, output_features=5120, bias=False, tp_size=8, reduce_results=True)
          (act_fn): SiluAndMul()
        )
        (post_attention_layernorm): RMSNorm(hidden_size=5120, eps=1e-06)
        (input_layernorm): Identity()
      )
    )
    (fc): Linear(in_features=10240, out_features=5120, bias=False)
  )
  (lm_head): ParallelLMHead(num_embeddings=18992, embedding_dim=5120, org_vocab_size=151936, num_embeddings_padded=151936, tp_size=8)
  (logits_processor): LogitsProcessor(vocab_size=151936, org_vocab_size=151936, scale=1.0, logits_as_input=False)
)
```
Quick start script:
```
nohup python -m vllm.entrypoints.openai.api_server \
    --trust-remote-code \
    --model=$model_path \
    --served-model-name auto \
    --host $host \
    --port $port \
    --tensor-parallel-size $tp \
    --data-parallel-size 1 \
    --no-enforce-eager \
    --max-num-seqs 256  \
    --max-num-batched-tokens 32768 \
    --gpu-memory-utilization 0.90 \
    --speculative_config '{"model": "modelpath", "num_speculative_tokens": 5, "method": "eagle3", "draft_tensor_parallel_size": 2}' \
    --compilation-config '{"cudagraph_capture_sizes": [6,12,24],"cudagraph_mode": "FULL_DECODE_ONLY"}' \
    > run.log &
```
The reason is that draft model shares same communication group with main model. We reuse function `patch_tensor_parallel_group` to deal with the problem.
Possibly related to issue #31345

## Test Plan
Test in server mode, print acceptance length to evaluate draft model accuracy with different tp size.

## Test Result
test in deepseek-v3 with self-data
```
(APIServer pid=79877) INFO 01-13 15:48:00 [metrics.py:100] SpecDecoding metrics: Mean acceptance length: 3.20, Accepted throughput: 31.70 tokens/s, Drafted throughput: 71.99 tokens/s, Accepted: 317 tokens, Drafted: 720 tokens, Per-position acceptance rate: 0.729, 0.549, 0.382, 0.312, 0.229, Avg Draft acceptance rate: 44.0%
```
test in qwen3-235b with self-data
```
(APIServer pid=85341) INFO 01-13 16:29:06 [metrics.py:100] SpecDecoding metrics: Mean acceptance length: 4.12, Accepted throughput: 23.40 tokens/s, Drafted throughput: 52.49 tokens/s, Accepted: 234 tokens, Drafted: 525 tokens, Per-position acceptance rate: 0.787, 0.653, 0.480, 0.387, 0.293, 0.293, 0.227, Avg Draft acceptance rate: 44.6%
```
---

> [!NOTE]
> Adds end-to-end support for draft models using a different tensor-parallel (TP) size than the target model.
> 
> - Relax `SpeculativeConfig._verify_and_get_draft_tp` validation so `draft_tensor_parallel_size` is no longer restricted to `{1, target TP}`
> - In `GPUModelRunner`, initialize a dedicated draft TP process group and run drafter ops under `patch_tensor_parallel_group(...)` (for load, propose, dummy_run)
> - In `EagleProposer`:
>   - Store original `tp_group` and implement `_reshard_draft_tp_tensor(...)` that `all_gather`s shards and slices to match the draft TP partition
>   - When sharing `embed_tokens`/`lm_head`, if shapes differ, reshard target weights to the draft TP shard; otherwise reuse modules directly
>   - Minor import consolidation (`get_pp_group`, `get_tp_group`) and small logging adjustments
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 948067ad6de220509d0d1c365ba6a9835b282b22. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Review by ywang96 — COMMENTED** [review]
I'm not sure if supporting a draft model parallelism other than replicating or the same TP as main model is something we want - what's the reason behind doing this?

cc @benchislett

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -1299,6 +1322,40 @@ def _pad_batch_across_dp(
...
+        Support different tp size between main model and draft model.
+        """
+        from vllm.distributed import get_tensor_model_parallel_rank
+
+        tp_rank = get_tensor_model_parallel_rank()
+        sd = target_model.state_dict()
+        dst_tensor = None
+        for name, param in self.model.named_parameters():
+            if target_name in name:
+                src_tensor = sd[name]
```
[the remark this answers is not in the document]
Comment (stormchasingg) [reply vllm/v1/spec_decode/eagle.py:1349]: Key error should be reported.

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -1299,6 +1322,40 @@ def _pad_batch_across_dp(
...
+                full_tensor = torch.empty(
+                    full_tensor_shape, dtype=src_tensor.dtype, device=src_tensor.device
+                )
+                torch.distributed.all_gather_into_tensor(
+                    full_tensor, src_tensor, group=self.tp_group
+                )
+                dst_tensor = full_tensor[
+                    tp_rank * param.shape[0] : (tp_rank + 1) * param.shape[0]
+                ]
+                logger.info(
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py]: This can be a debug-level log

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -1299,6 +1322,40 @@ def _pad_batch_across_dp(
             num_tokens_dp_padded = int(num_toks_across_dp[self.dp_rank].item())
         return num_tokens_dp_padded, num_toks_across_dp
 
+    def _reshard_draft_tp_tensor(self, target_model, target_name):
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py]: add type annotations

File: vllm/v1/spec_decode/eagle.py [hunk 4]
```diff
@@ -1299,6 +1322,40 @@ def _pad_batch_across_dp(
...
+        """
+        Support different tp size between main model and draft model.
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py:1339]: Can you add an early-return for this function if TP size is the same between main and draft model?

**Review by benchislett — CHANGES_REQUESTED** [review]
Left some comments, but overall I'm fine with the feature if it's useful. 

Can you also make sure that it will work with DP (if it is supported in the first place) and/or EP without any problems?

File: vllm/v1/spec_decode/eagle.py (hunk 3 above)
Comment (stormchasingg) [reply vllm/v1/spec_decode/eagle.py]: Fixed.

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
Comment (stormchasingg) [reply vllm/v1/spec_decode/eagle.py]: Fixed.

**Comment by stormchasingg** [comment]
@benchislett Thanks to your comments. We have tested 'dp + ep' main model with 'tp' draft model, such as deepseek-v3. But other parallelism is not supported in draft model, since we don't have a moe draft model. By the way, it needs a fully sharded manager to deal with other parallelism in draft model if the draft model becomes larger.

File: vllm/v1/spec_decode/eagle.py (hunk 4 above)
Comment (stormchasingg) [reply vllm/v1/spec_decode/eagle.py:1339]: Fixed.

**Comment by hknx001** [comment]
Hi @stormchasingg,

We've been hitting exactly the issue this PR addresses on a 4× DGX Spark / GB10 cluster running GLM 5.2 (744B/40B MoE, GlmMoeDsa architecture) with MTP=3 via vLLM 0.23.1rc1.dev190 (TP=4). GLM 5.2's MTP head goes through vLLM's `deepseek_v4` MTP module, same code path the fix touches, even though the model itself isn't a DeepSeek model. When we set `draft_tensor_parallel_size: 1`, the field is silently orphaned, EagleProposer in V1 doesn't honor it and the MTP layer ends up running at the target's TP=4, which costs roughly one full all-reduce per speculative step and largely cancels out the MTP speedup. From what we've read of this PR and the underlying issue (#31345), the approach you took is the right shape for the fix.
  
We can apply this PR to our chthonic-based vLLM image, run it on the actual 4×GB10 cluster with GLM 5.2 + MTP=3 as the workload, and report measured before/after numbers (decode tok/s, MTP acceptance rates, GPU memory at TP=4 vs TP=1 for the draft). That's the test coverage the reviewer asks flagged as missing.
  
Is the PR still active? Happy to:
  - Just run validation and report results back (no PR changes from us)
  - Pick up the reviewer asks (`all_gather` communicator, parameterized
    reshape, type annotations) and push commits if you're not on it
  - Whatever shape of contribution is most useful
  
If you've moved on and would rather we open a fresh PR building on your approach (with attribution), that works too. Mostly want to make sure the work doesn't get lost to stale-bot.
  
Thanks for the work so far.
